src/Analysis/Ast/Impl/Analyzer/Evaluation/ExpressionEval.Generics.cs

```diff
@@ -59,40 +63,62 @@ internal sealed partial class ExpressionEval {
+            if (genericTypeArgs.Length == 0) {
+                return false;
             }
-            if (specificTypes.Length > 0 && specificTypes.Length != indices.Count) {
-                // TODO: report that arguments are not specific types or are not declared.
+
+            // All arguments to Generic must be type parameters
+            // e.g. Generic[T, str] throws a runtime error
+            if (genericTypeArgs.Length != indices.Count) {
+                ReportDiagnostics(Module.Uri, new DiagnosticsEntry(
+                    Resources.GenericArgumentsNotAllTypeParameters,
+                    GetLocation(expr).Span,
+                    ErrorCodes.GenericArguments,
+                    Severity.Error,
+                    DiagnosticSource.Analysis));
+                return false;
+            }
+
+            // All arguments to Generic must be distinct
+            if (genericTypeArgs.Distinct().Count() != genericTypeArgs.Length) {
+                ReportDiagnostics(Module.Uri, new DiagnosticsEntry(
+                   Resources.GenericArgumentsNotAllUnique,
+                   GetLocation(expr).Span,
+                   ErrorCodes.GenericArguments,
+                   Severity.Error,
+                   DiagnosticSource.Analysis));
+                return false;
             }
 
+            return true;
+        }
+
+        /// <summary>
+        /// Given generic type and list of arguments in the expression like
+        /// Generic[T1, T2, ...] or List[str] creates generic class base
+        /// (if the former) on specific type (if the latter).
+        /// </summary>
+        private IMember CreateSpecificTypeFromIndex(IGenericType gt, IReadOnlyList<IMember> args, Expression expr) {
+            var genericTypeArgs = args.OfType<IGenericTypeDefinition>().ToArray();
+
             if (gt.Name.EqualsOrdinal("Generic")) {
-                // Generic[T1, T2, ...] expression. Create generic base for the class.
-                if (genericTypeArgs.Length > 0) {
-                    return new GenericClassParameter(genericTypeArgs, Module);
-                } else {
-                    // TODO: report too few type arguments for Generic[].
+                if (!GenericClassParameterValid(genericTypeArgs, args, expr)) {
                     return UnknownType;
                 }
+
+                // Generic[T1, T2, ...] expression. Create generic base for the class.
+                return new GenericClassParameter(genericTypeArgs, Module);
             }
 
             // For other types just use supplied arguments
-            if (indices.Count > 0) {
-                return gt.CreateSpecificType(new ArgumentSet(indices));
+            if (args.Count > 0) {
+                return gt.CreateSpecificType(new ArgumentSet(args));
             }
-            // TODO: report too few type arguments for the generic expression.
```

Comment:
Same here, when there are no arguments to a generic structure like List[str] being List[] that is a syntax error.

Reply:
`List` should be handled in its specialization. See `CreateListType` and others

Reply:
Ok I've found it, thanks!